backend/models/sequenceActivities.js

```diff
@@ -0,0 +1,337 @@
+		parents = [];
+		(currentDescendants || []).forEach(({_id}) => {
+			parents.push(_id);
+			ids.push(_id);
+		});
+	} while(currentDescendants.length);
+
+	return ids;
+};
+
+const addToActivityTree = (activity, parent, treeFile, treeFileDictionary) => {
+	// The treeFileDictionary is being use for quick access to the activity,
+	// in particular to add the subactivities to its parents
+	const parentId = parent ? utils.uuidToString(parent._id || parent.id) : null ;
+	const id = utils.uuidToString(activity._id || activity.id);
+
+	if (parent && !treeFileDictionary[parentId]) {
+		treeFileDictionary[parentId] = simplifyActivity(parent);
+	}
+
+	if (!treeFileDictionary[id]) {
+		treeFileDictionary[id] = simplifyActivity(activity);
+	}
+
+	if (parent) {
+		if (!treeFileDictionary[parentId].subActivities) {
+			treeFileDictionary[parentId].subActivities = [];
+		}
+
+		treeFileDictionary[parentId].subActivities.push(treeFileDictionary[id]);
+	} else {
+		treeFile.push(treeFileDictionary[id]);
+	}
+};
+
+const createActivitiesTree = async(account, model, sequenceId) => {
+	const foundActivities = await db.find(account, model + ".activities", {sequenceId: utils.stringToUUID(sequenceId)}, { metadata: 0, resources: 0 });
+
+	const activities = [];
+	const activitiesDictionary = {};
+
+	foundActivities.forEach((activity) => {
+		activitiesDictionary[utils.uuidToString(activity._id)] = simplifyActivity(activity);
+	});
+
+	foundActivities.forEach((activity) => {
+		const parent = activity.parent ? activitiesDictionary[utils.uuidToString(activity.parent)] : null;
+		addToActivityTree(activity, parent, activities, activitiesDictionary);
+	});
+
+	return { activities };
+};
+
+// This function is used for validation the receive data from the request,
+// add an activity to a plain array and create the treefile for saving afterwards
+const addToActivityListAndCreateTreeFile = (activitiesList, sequenceIdUUID, treeFile, createFile) => {
+	const treeFileDictionary = {};
+
+	let id = null;
+	const idsDictionary = {};
+
+	return (activity, parent) => {
+		// TODO: test that _id is a bson object
+		if (!activitySchema.isValidSync(activity, { strict: true })) {
+			throw responseCodes.INVALID_ARGUMENTS;
+		}
+
+		if (parent && !parent._id) {
+			id = utils.stringToUUID(nodeuuid());
+
+			while (idsDictionary[id]) { // if the id is already there, create a new one
+				id = utils.stringToUUID(nodeuuid());
+			}
+
+			idsDictionary[id] = true;
+
+			parent._id = id ;
+		}
+
+		if (!activity._id) {
+			id = utils.stringToUUID(nodeuuid());
+
+			while (idsDictionary[id]) { // if the id is already there, create a new one
+				id = utils.stringToUUID(nodeuuid());
+			}
+
+			idsDictionary[id] = true;
+			activity._id =  id;
+		}
+
+		const plainActivity = pick(activity,"_id", "name", "startDate", "endDate", "data", "resources");
+		plainActivity.sequenceId = sequenceIdUUID;
+
+		if (parent) {
+			plainActivity.parent = parent._id;
```

Comment:
we can just pick it from the activity if we do the rewrite in `traverseActivities`